taskgraph/Task.py

```diff
@@ -767,8 +766,12 @@ def join(self, timeout=None):
 
         """
         LOGGER.debug("joining taskgraph")
-        if self._n_workers < 0 or self._terminated:
+        if self._n_workers < 0:

```

Comment:
Sorry, I'm a little confused trying to parse out how this working now. If `self._terminated` is `true`, the below `try/except` will raise the exception for the 2nd time? What part of that `try` do we expect to fail which will raise the exception?

Reply:
> What part of that try do we expect to fail which will raise the exception?
It's the `task.join()` line.  Any exceptions raised within a task are then re-raised by `task.join()`.  So then when we do a `graph.join()`, any tasks with exceptions will have their errors propagated.

> If self._terminated is true, the below try/except will raise the exception for the 2nd time?
Well, it depends!  Let's say we have a task `E` that raises `ValueError`.

If we call `E.join()`, then that will raise `ValueError`.  When we call `graph.join()` at the end of the pipeline, `graph.join()` will _also_ raise `ValueError`.

But if we _don't_ call `E.join()` (which is what happens in most InVEST models), `E`'s exception will terminate the graph and log the exception, BUT the exception will never be raised.  This is the behavior that's changing in this PR ... if the graph is terminated, we still raise the error that caused it.

This second case is why we were getting the flaky `ZeroDivisionError` ... the task was starting up so fast that by the time we got to `graph.join()`, the error was logged but wouldn't be raised.  `graph.join()` was only raising the error if and only if there were any tasks remaining by the time `graph.join()` is called.

Reply:
>If we call E.join(), then that will raise ValueError. When we call graph.join() at the end of the pipeline, graph.join() will also raise ValueError.

>But if we don't call E.join() (which is what happens in most InVEST models), E's exception will terminate the graph and log the exception, BUT the exception will never be raised. This is the behavior that's changing in this PR ... if the graph is terminated, we still raise the error that caused it.

Brilliant, thanks for clarifying, that helped a lot.